File: C++/Problems/Problems/problem_2592.cpp
#include "problem_2592.h"

#include <algorithm>

bool problem_2592::test() {
    std::vector nums { 1, 3, 5, 2, 1, 3, 1 };

    int expected = 4;

    auto result = maximizeGreatness(nums);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_2592::maximizeGreatness(std::vector<int> &nums) {
    int n = nums.size();
    if (n < 2)
        return 0;
    std::sort(nums.begin(), nums.end());
    int greatness = 0;
    int start = 0;
    for (int i = 1; i < n; ++i)
    {
        if (nums[i] != nums[i - 1])
        {
            int l = i - start;
            greatness = std::min(greatness + l, start);
            start = i;
        }
    }

    int fl = n - start;
    greatness = std::min(greatness + fl, start);

    return greatness;
}

File: Python/Problems/Problems/problem2592.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem2592(Problem):
    def test(self) -> bool:
        nums = [1, 3, 5, 2, 1, 3, 1]

        expected = 4

        result = self.maximizeGreatness(nums)

        print(result)

        return result == expected

    def maximizeGreatness(self, nums: List[int]) -> int:
        n = len(nums)
        if n < 2:
            return 0
        nums.sort()
        greatness = 0
        start = 0
        for i in range(1, n):
            if nums[i] != nums[i - 1]:
                l = i - start
                greatness = min(greatness + l, start)
                start = i

        fl = n - start
        greatness = min(greatness + fl, start)

        return greatness